### PR TITLE
Remove DecoderState Class (for future Onnx export)

### PR DESCRIPTION
**Pull request by vince62s**
This is necessary for future Onnx export.

This "simplifies" the way we maintain the decoder state within the Decoder itself.

### CODE REVIEW

File: onmt/decoders/decoder.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ class RNNDecoderBase(nn.Module):
           G[Decoder State]
           H[Decoder State]
           I[Outputs]
-          F[Memory_Bank]
+          F[attn_context]
```
Comment (guillaumekln) [inline onmt/decoders/decoder.py]: Not sure why this change of terminology is part of this PR. In addition to that, I don't think it's a good change: context usally refers to the output of the attention layer and not the output of the encoder. I strongly suggest reverting this renaming from this PR.

File: onmt/decoders/decoder.py [hunk 2]
```diff
@@ -168,12 +169,28 @@ def _fix_enc_hidden(hidden):
...
+                                          for enc_hid in encoder_final])
         else:  # GRU
-            return RNNDecoderState(self.hidden_size,
-                                   _fix_enc_hidden(encoder_final))
+            self.state["hidden"] = (_fix_enc_hidden(encoder_final), )
+
+        # Init the input feed.
+        batch_size = self.state["hidden"][0].size(1)
+        h_size = (batch_size, self.hidden_size)
+        self.state["input_feed"] = \
```
Comment (guillaumekln) [inline onmt/decoders/decoder.py]: The input feeding logic should only be part of `InputFeedRNNDecoder`. You can add overrides in this class, call the base methods and add the input feed state.

File: onmt/decoders/decoder.py [hunk 3]
```diff
@@ -168,12 +169,28 @@ def _fix_enc_hidden(hidden):
...
+            self.state["hidden"][0].data.new(*h_size).zero_().unsqueeze(0)
+        self.state["coverage"] = None
+
+    def map_batch_fn(self, fn):
```
Comment (guillaumekln) [inline onmt/decoders/decoder.py]: I suggest naming this method `map_state` for clarity.

File: onmt/decoders/decoder.py [hunk 4]
```diff
@@ -168,12 +169,28 @@ def _fix_enc_hidden(hidden):
...
+        self.state["hidden"] = tuple(map(lambda x: fn(x, 1),
+                                         self.state["hidden"]))
+        self.state["input_feed"] = fn(self.state["input_feed"], 1)
+
+    def detach(self):
```
Comment (guillaumekln) [inline onmt/decoders/decoder.py]: I suggest naming this method `detach_state`.

Additionally, this method can be implemented using `map_batch_fn` (or `map_state`). This way, only the base class needs to define it (an abstract base class `Decoder` would be helpful here).

```python
self.map_state(lambda x, _: x.detach())
```

File: onmt/decoders/decoder.py [hunk 5]
```diff
@@ -315,20 +332,20 @@ def _run_forward_pass(self, tgt, memory_bank, state, memory_lengths=None):
...
+            if self.state["coverage"] is not None else None
 
         # Input feed concatenates hidden state with
         # input at every time step.
         for _, emb_t in enumerate(emb.split(1)):
             emb_t = emb_t.squeeze(0)
             decoder_input = torch.cat([emb_t, input_feed], 1)
-
-            rnn_output, hidden = self.rnn(decoder_input, hidden)
+            print(decoder_input.size(), dec_state[0].size())
```
Comment (guillaumekln) [inline onmt/decoders/decoder.py]: Remove this print.

File: onmt/decoders/cnn_decoder.py [hunk 6]
```diff
@@ -102,49 +101,37 @@ def forward(self, tgt, memory_bank, state, memory_lengths=None, step=None):
...
         # Update the state.
-        state.update_state(tgt)
+        self.update_state(tgt)
+        # TODO change the way attns is returned dict => list or tuple (onnx)
+        return dec_outs, attns
 
-        return outputs, state, attns
-
-    def init_decoder_state(self, _, memory_bank, enc_hidden, with_cache=False):
+    def init_decoder_state(self, _, attn_context, enc_hidden,
```
Comment (guillaumekln) [inline onmt/decoders/cnn_decoder.py]: For consistency with `update_state`, I suggest renaming this to `init_state`.

File: onmt/decoders/decoder.py (hunk 4 above)
Comment (vince62s) [reply onmt/decoders/decoder.py]: Will do this later when creating an abstract base class.
would like to do some other renaming first.

File: onmt/decoders/decoder.py (hunk 2 above)
Comment (guillaumekln) [reply onmt/decoders/decoder.py]: Let's make another PR after this one that removes `update_state`. It does not appear useful over directly updating the state fields.

Comment (vince62s) [reply onmt/decoders/decoder.py]: yes, correct.

**Comment by vince62s** [comment]
Agree 100% for points 2 to 4.
Not sure to understand the context of 1. From a general poitn of view or specifically in the code (eg: when running decode with gold score ?)

**Comment by guillaumekln** [comment]
Both. Target scoring is an example of decoder reuse and more generally, the API should make this clear.

By the way, #1021 is incompatible with this PR as it was designed before it.
